File: manual/code/generate_strategies/IR(CoT)/python/python/function_pandas_pivot_table.py
import pandas as pd


class CreatePivotTable:

    def create_pivot_table(self, date1=None, date2=None, category1=None, category2=None, value1=None, value2=None):
        if date1 is None or date2 is None or category1 is None or category2 is None or value1 is None or value2 is None:
            date1 = "2023-01-01"
            date2 = "2023-01-02"
            category1 = "A"
            category2 = "B"
            value1 = 1
            value2 = 2

        data = {
            'Date': [date1, date2],
            'Category': [category1, category2],
            'Values': [value1, value2]
        }
        df = pd.DataFrame(data)

        pivot_df = pd.pivot_table(df, index='Date', columns='Category', values='Values', aggfunc='sum', fill_value=0)

        result = f"Date: {pivot_df.index[0]}, Category A: {pivot_df.iloc[0]['A']}, Category B: {pivot_df.iloc[0]['B']}"
        return result

File: manual/code/generate_strategies/IR(CoT)/cpp/cpp/function_pandas_pivot_table.cpp
#include <string>
#include <sstream>
#include <optional>
#include <xtensor/xarray.hpp>
#include <xtensor/xview.hpp>
#include <xtensor/xadapt.hpp>

class CreatePivotTable {
public:


    template<typename T>
    std::string create_pivot_table(
        const std::optional<std::string>& date1 = std::nullopt,
        const std::optional<std::string>& date2 = std::nullopt,
        const std::optional<std::string>& category1 = std::nullopt,
        const std::optional<std::string>& category2 = std::nullopt,
        const std::optional<T>& value1 = std::nullopt,
        const std::optional<T>& value2 = std::nullopt)
    {
        
        std::string d1 = date1.value_or("2023-01-01");
        std::string c1 = category1.value_or("A");
        std::string c2 = category2.value_or("B");
        T v1 = value1.value_or(static_cast<T>(1));
        T v2 = static_cast<T>(0);  

        
        if (value2.has_value()) {
            v2 = value2.value();
        }

        
        xt::xarray<T> values = {v1, v2};
        std::vector<std::string> categories = {c1, c2};
        
        
        std::ostringstream oss;
        oss << "Date: " << d1 
            << ", Category A: " << (c1 == "A" ? values(0) : 0)
            << ", Category B: " << (c2 == "B" ? values(1) : 0);

        return oss.str();
    }
};